File: 1_flow/run_decision_tree.R
#----------------------------------------------
# ライブラリ
#----------------------------------------------
library(rpart)
library(rpart.plot)

#----------------------------------------------
# ディレクトリ・ファイル設定（親ディレクトリ基準）
#----------------------------------------------
current_path <- getwd()
parent_path <- dirname(current_path)

INPUT_FOLDER  <- "2_data"
OUTPUT_FOLDER <- "3_output"
INPUT_FILE    <- "sample.csv"

input_path  <- file.path(parent_path, INPUT_FOLDER, INPUT_FILE)
output_path <- file.path(parent_path, OUTPUT_FOLDER)

if (!dir.exists(output_path)) dir.create(output_path)

# 出力ファイル名
save_name     <- file.path(output_path, "R_decision_tree.png")    # 比率あり
non_save_name <- file.path(output_path, "R_decision_tree_N.png")  # 人数のみ

#----------------------------------------------
# CSV読み込み（文字コード自動判定）
#----------------------------------------------
# まずUTF-8で読み込み、失敗したらCP932（Shift-JIS）で
df <- tryCatch({
  read.csv(input_path, fileEncoding = "UTF-8")
}, error = function(e) {
  read.csv(input_path, fileEncoding = "CP932")
})

#----------------------------------------------
# 除外カラム設定・目的変数
#----------------------------------------------
ID <- "ID"
cluster <- "cluster"
target_col <- "car"

x_df <- df[, setdiff(names(df), c(ID, cluster, target_col))]
y_df <- df[[target_col]]

# rpart用に結合
data_all <- data.frame(y_df, x_df)

#----------------------------------------------
# 訓練・テストデータ分割（75% / 25%）
#----------------------------------------------
set.seed(0)  # 再現性確保
train_idx <- sample(seq_len(nrow(data_all)), size = 0.75 * nrow(data_all))
train_data <- data_all[train_idx, ]
test_data  <- data_all[-train_idx, ]



#----------------------------------------------
# 決定木モデル構築（max_depth=3）
#----------------------------------------------
model <- rpart(y_df ~ ., 
               data = train_data, 
               method = "class", 
               parms = list(split = "gini"),
               control = rpart.control(maxdepth = 3))

# 正解率計算
train_pred <- predict(model, train_data, type = "class")
train_acc <- mean(train_pred == train_data$y_df)

test_pred <- predict(model, test_data, type = "class")
test_acc <- mean(test_pred == test_data$y_df)

cat(sprintf("学習データ数: %d / テストデータ数: %d\n", nrow(train_data), nrow(test_data)))
cat(sprintf("正解率（train）: %.3f / 正解率（test）: %.3f\n", train_acc, test_acc))

#----------------------------------------------
# 決定木作成（全データ, maxdepth=6）
#----------------------------------------------
model_full <- rpart(y_df ~ ., 
                    data = data_all, 
                    method = "class", 
                    parms = list(split = "gini"),
                    control = rpart.control(maxdepth = 6))

#----------------------------------------------
# 特徴量
#----------------------------------------------
print(model$variable.importance)

#----------------------------------------------
# 決定木プロット（比率あり） & PNG保存
#----------------------------------------------
png(save_name, width = 1000, height = 800)
rpart.plot(model_full, type = 2, extra = 104, fallen.leaves = TRUE, main = "決定木(比率)")
dev.off()

#----------------------------------------------
# 決定木プロット（人数のみ） & PNG保存
#----------------------------------------------
png(non_save_name, width = 1000, height = 800)
rpart.plot(model_full, type = 2, extra = 1, fallen.leaves = TRUE, main = "決定木（人数のみ）")
dev.off()

cat("処理完了\n")

File: 1_flow/run_decision_tree.py
# ---
# jupyter:
#   jupytext:
#     formats: py:percent
#     text_representation:
#       extension: .py
#       format_name: percent
#       format_version: '1.3'
#       jupytext_version: 1.17.1
#   kernelspec:
#     display_name: Python 3 (ipykernel)
#     language: python
#     name: python3
# ---

# %%
import os
import sys
import subprocess
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier, plot_tree
import matplotlib.pyplot as plt
import japanize_matplotlib  # matplotlibで日本語ラベルを扱うためのライブラリ

# ディレクトリ・ファイル設定
INPUT_FOLDER = '2_data'
INPUT_FILE = 'sample.csv'
OUTPUT_FOLDER = '3_output'

save_name  = 'decision_tree.png'       # 比率あり
non_save_name  = 'decision_tree_N.png' # 比率なし

# 除外カラム設定
ID = 'ID'
cluster = 'cluster'

#目的変数
target_col = 'car' 

# パス設定
current_path = os.getcwd()
parent_path = os.path.dirname(current_path)

input_path = os.path.join(parent_path, INPUT_FOLDER, INPUT_FILE)
output_path = os.path.join(parent_path, OUTPUT_FOLDER)
os.makedirs(output_path, exist_ok=True)  # フォルダがなければ作成

save_path = os.path.join(output_path, save_name)
Non_save_path = os.path.join(output_path, non_save_name)

# CSV読み込み（文字コード自動判定）
try:
    df = pd.read_csv(input_path, encoding="utf-8")
except UnicodeDecodeError:
    df = pd.read_csv(input_path, encoding="cp932")

# %%
# 説明変数と目的変数に分ける
X_df = df.drop([ID, cluster, target_col], axis=1)  # cluster列は事前に作成されていると仮定
y_df = df[target_col]  # 目的変数（car）

# 訓練・テストデータに分割
X_train, X_test, y_train, y_test = train_test_split(X_df, y_df, random_state=0)
print(f"学習データ数: {len(X_train)} / テストデータ数: {len(X_test)}")

# 決定木モデル構築
clf = DecisionTreeClassifier(criterion='gini', max_depth=3, random_state=0)
model = clf.fit(X_train, y_train)

print(
    f"正解率（train）: {model.score(X_train, y_train):.3f} / "
    f"正解率（test）: {model.score(X_test, y_test):.3f}"
)


# %%
def add_class_legend(legend_text, x=0.02, y=0.98):
    """決定木のクラス凡例をFigureに追加"""
    plt.gcf().text(
        x, y,
        legend_text,
        fontsize=13,
        va="top",
        ha="left",
        bbox=dict(
            boxstyle="round,pad=0.5",
            fc="lightyellow",
            ec="gray",
            lw=1
        )
    )

# クラスインデックス対応表
legend_text = "Class index mapping:\n" + "\n".join(
    [f"{idx}: {label}" for idx, label in enumerate(model.classes_)]
)

# 比率なし
plt.figure(figsize=(20,10))
plot_tree(model, feature_names=X_df.columns,
          class_names=model.classes_.astype(str),
          filled=True, max_depth=6)
add_class_legend(legend_text)
plt.savefig(Non_save_path, dpi=300)
plt.show()

# 比率あり
plt.figure(figsize=(20,10))
plot_tree(model, feature_names=X_df.columns,
          class_names=model.classes_.astype(str),
          filled=True, max_depth=6, proportion=True)
add_class_legend(legend_text)
plt.savefig(save_path, dpi=300)
plt.show()

# 特徴量
for name, importance in zip(X_df.columns, model.feature_importances_):
    print(f"{name}: {importance:.3f}")

# OS別で出力フォルダを開く
if sys.platform.startswith('win'):
    os.startfile(output_path)
elif sys.platform.startswith('darwin'):
    subprocess.run(['open', output_path])
else:
    subprocess.run(['xdg-open', output_path])

print('処理完了')
